File: Sources/PrivateWhisper/TextInjector.swift
import AppKit
import Carbon.HIToolbox

enum InjectionOutcome {
    case injected
    /// Text was not injected; the caller should show it in the HUD instead.
    case needsHUD(reason: String)
}

/// Injects text at the cursor of the frontmost app: pasteboard + synthetic
/// Cmd+V, then restores the previous pasteboard contents (PRD §4.1-D).
enum TextInjector {
    /// Roles that clearly cannot receive pasted text. Anything unknown is
    /// treated as paste-able — a stray Cmd+V into a non-text target is a no-op,
    /// while a false HUD would interrupt the flow. Heuristic, documented in
    /// DECISIONS.md.
    private static let nonTextRoles: Set<String> = [
        kAXButtonRole, kAXMenuItemRole, kAXMenuRole, kAXImageRole,
        kAXCheckBoxRole, kAXRadioButtonRole, kAXSliderRole,
    ]

    @MainActor
    static func inject(_ text: String) -> InjectionOutcome {
        guard AXIsProcessTrusted() else {
            return .needsHUD(reason: "Accessibility permission not granted")
        }
        if IsSecureEventInputEnabled() {
            return .needsHUD(reason: "A secure input field is active (e.g. a password field)")
        }
        switch focusedElementAssessment() {
        case .noFocus:
            // The systemwide AX probe fails in many apps (Electron, browsers,
            // web views) even when a text field clearly has focus — "cannot
            // determine" must not block injection. A Cmd+V with nothing
            // focused is a harmless no-op, and the menu bar keeps a
            // "Copy Last Dictation" fallback.
            dlog("AX focused-element probe inconclusive — injecting anyway")
        case .nonText(let role):
            return .needsHUD(reason: "Focused element (\(role)) is not a text field")
        case .textLikeOrUnknown:
            break
        }

        paste(text)
        return .injected
    }

    private enum FocusAssessment {
        case noFocus
        case nonText(role: String)
        case textLikeOrUnknown
    }

    private static func focusedElementAssessment() -> FocusAssessment {
        let systemWide = AXUIElementCreateSystemWide()
        var focused: CFTypeRef?
        let err = AXUIElementCopyAttributeValue(
            systemWide, kAXFocusedUIElementAttribute as CFString, &focused)
        guard err == .success, let element = focused else { return .noFocus }

        var roleValue: CFTypeRef?
        AXUIElementCopyAttributeValue(
            element as! AXUIElement, kAXRoleAttribute as CFString, &roleValue)
        if let role = roleValue as? String, nonTextRoles.contains(role) {
            return .nonText(role: role)
        }
        return .textLikeOrUnknown
    }

    /// Monotonic generation so only the latest scheduled restore runs when
    /// dictations overlap within the restore window.
    @MainActor private static var restoreGeneration = 0

    @MainActor
    private static func paste(_ text: String) {
        let pasteboard = NSPasteboard.general

        // Snapshot every item/type so we can restore rich content afterwards.
        let snapshot: [[NSPasteboard.PasteboardType: Data]] =
            (pasteboard.pasteboardItems ?? []).map { item in
                var entry: [NSPasteboard.PasteboardType: Data] = [:]
                for type in item.types {
                    if let data = item.data(forType: type) { entry[type] = data }
                }
                return entry
            }

        pasteboard.clearContents()
        pasteboard.setString(text, forType: .string)
        let expectedChangeCount = pasteboard.changeCount
        restoreGeneration += 1
        let generation = restoreGeneration

        postCmdV()

        // Give the target app time to read the pasteboard before restoring.
        // Skip the restore if anyone (the user copying, a newer dictation)
        // wrote to the pasteboard in the meantime — never clobber newer content.
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) {
            guard generation == restoreGeneration,
                  pasteboard.changeCount == expectedChangeCount else { return }
            pasteboard.clearContents()
            let items = snapshot.map { entry -> NSPasteboardItem in
                let item = NSPasteboardItem()
                for (type, data) in entry { item.setData(data, forType: type) }
                return item
            }
            if !items.isEmpty { pasteboard.writeObjects(items) }
        }
    }

    private static func postCmdV() {
        let source = CGEventSource(stateID: .combinedSessionState)
        let vKey: CGKeyCode = 9
        guard let down = CGEvent(keyboardEventSource: source, virtualKey: vKey, keyDown: true),
              let up = CGEvent(keyboardEventSource: source, virtualKey: vKey, keyDown: false)
        else { return }
        down.flags = .maskCommand
        up.flags = .maskCommand
        down.post(tap: .cghidEventTap)
        up.post(tap: .cghidEventTap)
    }
}

File: windows/src/PrivateWhisper/TextInjector.cs
using DataFormats = System.Windows.DataFormats;
using DataObject = System.Windows.DataObject;
using IDataObject = System.Windows.IDataObject;
using Clipboard = System.Windows.Clipboard;
using System.Runtime.InteropServices;
using System.Windows;

namespace PrivateWhisper;

public enum InjectionKind
{
    Injected,
    /// <summary>Text was not injected; the caller should surface it instead.</summary>
    NeedsOverlay,
}

public readonly struct InjectionOutcome
{
    public InjectionKind Kind { get; }
    public string? Reason { get; }

    private InjectionOutcome(InjectionKind kind, string? reason)
    {
        Kind = kind;
        Reason = reason;
    }

    public static InjectionOutcome Injected() => new(InjectionKind.Injected, null);
    public static InjectionOutcome NeedsOverlay(string reason) => new(InjectionKind.NeedsOverlay, reason);
}

/// <summary>
/// Injects text at the cursor of the foreground app: clipboard + synthetic
/// Ctrl+V, then restores the previous clipboard contents (PRD §4.1-D).
/// The restore is guarded by GetClipboardSequenceNumber — if anyone (the user
/// copying, a newer dictation) wrote to the clipboard in the meantime, the
/// restore is skipped so newer content is never clobbered. Windows analog of
/// the Mac changeCount guard. Must be called on the UI (STA) thread.
/// </summary>
public static class TextInjector
{
    /// <summary>Monotonic generation so only the latest scheduled restore runs
    /// when dictations overlap within the restore window.</summary>
    private static int restoreGeneration;

    public static InjectionOutcome Inject(string text)
    {
        DataObject? snapshot = SnapshotClipboard();

        if (!TrySetClipboardText(text))
        {
            return InjectionOutcome.NeedsOverlay("Clipboard is busy (another app holds it open)");
        }
        uint expectedSequence = NativeMethods.GetClipboardSequenceNumber();
        int generation = ++restoreGeneration;

        SyntheticKeyboard.SendCtrlV();

        // Give the target app time to read the clipboard before restoring.
        _ = RestoreLaterAsync(snapshot, expectedSequence, generation);
        return InjectionOutcome.Injected();
    }

    private static async Task RestoreLaterAsync(DataObject? snapshot, uint expectedSequence, int generation)
    {
        await Task.Delay(600); // resumes on the UI thread (dispatcher context)
        if (generation != restoreGeneration) return;
        if (NativeMethods.GetClipboardSequenceNumber() != expectedSequence) return;
        try
        {
            if (snapshot == null)
            {
                Clipboard.Clear();
            }
            else
            {
                Clipboard.SetDataObject(snapshot, true);
            }
        }
        catch (Exception ex)
        {
            Log.D("clipboard restore failed: " + ex.Message);
        }
    }

    /// <summary>Snapshot every readable format so rich content survives the
    /// round-trip. Individual formats that refuse to serialize are skipped.</summary>
    internal static DataObject? SnapshotClipboard()
    {
        try
        {
            IDataObject? source = Clipboard.GetDataObject();
            if (source == null) return null;
            var copy = new DataObject();
            bool any = false;
            foreach (string format in source.GetFormats(false))
            {
                try
                {
                    if (source.GetDataPresent(format, false))
                    {
                        object? data = source.GetData(format, false);
                        if (data != null)
                        {
                            copy.SetData(format, data);
                            any = true;
                        }
                    }
                }
                catch
                {
                    // Some formats throw on read (delayed rendering, dead owner) — skip.
                }
            }
            return any ? copy : null;
        }
        catch (Exception ex)
        {
            Log.D("clipboard snapshot failed: " + ex.Message);
            return null;
        }
    }

    internal static bool TrySetClipboardText(string text)
    {
        for (int attempt = 0; attempt < 5; attempt++)
        {
            try
            {
                Clipboard.SetDataObject(new DataObject(DataFormats.UnicodeText, text), true);
                return true;
            }
            catch (COMException)
            {
                Thread.Sleep(40); // clipboard briefly locked by another app
            }
            catch (Exception ex)
            {
                Log.D("clipboard set failed: " + ex.Message);
                return false;
            }
        }
        return false;
    }
}
